File: frontend-reshet/src/services/workload-security.ts
import { httpClient } from "./http"

export type ApprovalStatus = "pending" | "approved" | "rejected"

export interface PendingScopePolicy {
  policy_id: string
  principal_id: string
  requested_scopes: string[]
  status: ApprovalStatus
  version: number
  created_at: string
}

export interface ScopePolicyApprovalResult {
  status: "approved"
  principal_id: string
  policy_id: string
  approved_scopes: string[]
  approved_at: string
}

export interface ScopePolicyRejectResult {
  status: "rejected"
  principal_id: string
  policy_id: string
}

export interface ActionApprovalDecision {
  id: string
  tenant_id?: string
  subject_type: string
  subject_id: string
  action_scope: string
  status: ApprovalStatus
  decided_by?: string | null
  rationale?: string | null
  created_at?: string
  decided_at?: string | null
}

class WorkloadSecurityService {
  async listPendingPolicies(): Promise<PendingScopePolicy[]> {
    return httpClient.get("/admin/security/workloads/pending")
  }

  async approveScopePolicy(
    principalId: string,
    approvedScopes: string[]
  ): Promise<ScopePolicyApprovalResult> {
    return httpClient.post(`/admin/security/workloads/principals/${principalId}/approve`, {
      approved_scopes: approvedScopes,
    })
  }

  async rejectScopePolicy(principalId: string): Promise<ScopePolicyRejectResult> {
    return httpClient.post(`/admin/security/workloads/principals/${principalId}/reject`)
  }

  async listActionApprovals(filters?: {
    subject_type?: string
    subject_id?: string
    action_scope?: string
  }): Promise<ActionApprovalDecision[]> {
    const params = new URLSearchParams()
    if (filters?.subject_type) params.set("subject_type", filters.subject_type)
    if (filters?.subject_id) params.set("subject_id", filters.subject_id)
    if (filters?.action_scope) params.set("action_scope", filters.action_scope)
    const query = params.toString() ? `?${params.toString()}` : ""
    return httpClient.get(`/admin/security/workloads/approvals${query}`)
  }

  async decideActionApproval(input: {
    subject_type: string
    subject_id: string
    action_scope: string
    status: Extract<ApprovalStatus, "approved" | "rejected">
    rationale?: string
  }): Promise<ActionApprovalDecision> {
    return httpClient.post("/admin/security/workloads/approvals/decide", input)
  }
}

export const workloadSecurityService = new WorkloadSecurityService()

File: backend/talmudpedia_control_sdk/workload_security.py
from __future__ import annotations

from typing import Any, Dict, Optional

from .types import ResponseEnvelope


class WorkloadSecurityAPI:
    def __init__(self, client: Any) -> None:
        self._client = client

    def list_pending_scope_policies(self) -> ResponseEnvelope:
        return self._client.request("GET", "/admin/security/workloads/pending")

    def approve_scope_policy(self, principal_id: str, approved_scopes: list[str]) -> ResponseEnvelope:
        payload = {"approved_scopes": approved_scopes}
        return self._client.request(
            "POST",
            f"/admin/security/workloads/principals/{principal_id}/approve",
            json_body=payload,
            mutation=True,
        )

    def reject_scope_policy(self, principal_id: str) -> ResponseEnvelope:
        return self._client.request(
            "POST",
            f"/admin/security/workloads/principals/{principal_id}/reject",
            json_body={},
            mutation=True,
        )

    def list_action_approvals(
        self,
        *,
        subject_type: Optional[str] = None,
        subject_id: Optional[str] = None,
        action_scope: Optional[str] = None,
    ) -> ResponseEnvelope:
        params: Dict[str, Any] = {}
        if subject_type:
            params["subject_type"] = subject_type
        if subject_id:
            params["subject_id"] = subject_id
        if action_scope:
            params["action_scope"] = action_scope
        return self._client.request("GET", "/admin/security/workloads/approvals", params=params)

    def decide_action_approval(self, payload: Dict[str, Any]) -> ResponseEnvelope:
        return self._client.request(
            "POST",
            "/admin/security/workloads/approvals/decide",
            json_body=payload,
            mutation=True,
        )
